lib/libmbedtls/core/dh.c

```diff
@@ -60,22 +61,22 @@ TEE_Result crypto_acipher_gen_dh_key(struct dh_keypair *key,
 	dhm.len = crypto_bignum_num_bytes(key->p);
 
 	if (xbits == 0)
-		xbits = dhm.len;
+		xbytes = dhm.len;
 	else
-		xbits = xbits / 8;
+		xbytes = xbits / 8;
 
 	buf = malloc(dhm.len);
 	if (!buf) {
 		res = TEE_ERROR_OUT_OF_MEMORY;
 		goto out;
 	}
-	lmd_res = mbedtls_dhm_make_public(&dhm, (int)xbits, buf,
+	lmd_res = mbedtls_dhm_make_public(&dhm, (int)xbytes, buf,
 					  dhm.len, mbd_rand, NULL);
 	if (lmd_res != 0) {
 		FMSG("mbedtls_dhm_make_public err, return is 0x%x", -lmd_res);
 		res = TEE_ERROR_BAD_PARAMETERS;
 	} else {
-		crypto_bignum_bin2bn(buf, xbits / 8, key->y);
```

Comment:
the problem is here, where xbits is already in convert to bytes is once again divided by 8.  So for example, DH_2048 bits key should return a 256 byte public key buffer, yet here it will return only 32 bytes.

Reply:
Ouch, that's a rather serious bug, thanks for finding and send a patch. At some point we should try to write a test case in xtest for this.

`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`